grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -2361,6 +2361,9 @@ class StockMovementService {
         }
 
         Container pallet = (palletName) ? shipment.findOrCreatePallet(palletName) : null
+        if (pallet) {
+            pallet.save(flush: true)
+        }
```

Comment:
what was the exact problem here? 
Also, if you find the pallet, what's the reason to save it?

I'm asking, because we had a discussion that we shouldn't just throw `flush: true` everywhere if it is not necessary.

Reply:
The problem was that when creating a shipment item and both provided box and pallet do not exist in the database, it creates both at that moment. When creating a box, it references the newly created pallet as a parentContainer. Since pallet is not yet saved in the database, and the save of all entities only happens at the end of the transaction, box references a non yet existing pallet as a parentContainer causing the parentContainer field to be empty.
So to solve this problem we need to make sure that in the moment of creating a box a pallet already exists.

I am not sure why this exact issue became a problem in grails 3, I wasn't able to find anything in the docs refferecnig this change, do you maybe have an idea ?

Reply:
@kchelstowski I'd say `shipment.findOrCreatePallet(palletName)` is not saved properly, hence it is fine (see how this method looks currently in the code)

Reply:
Ok, I get it, but:

1. `flush: true` doesn't make the changes being visible immediately in the DB if you are in the service layer
2. calling it without flush should be fine
3. I still don't get why we'd want to call `save()` if the `findOrCreatePallet` finds one (not create) - unnecessary save happens 
4. if you do it for `pallet`, why don't you do it also for the `box` below? The same problem should be visible there
5. I think in Grails 1 it was working because of the OSIV - here, it's fine that you've added this `save()`, because it is a new entity, but I just wanted to point out that `flush` might be unnecessary here, and also that we might call the `save()` for no reason there, if the method finds the Container.

Reply:
Yeah, I'm a little weary of this solution. I think a save within the findOrCreatePallet() would be more palatable ... pardon the pun. If not then we should look into an alternative solution.